### ISSUE TITLE
Possibly migration to Plotly.js charting library

### ISSUE THREAD

**Issue by veggiesaurus** [CARTAvis/carta-frontend#586]
_(This is something I have been discussing informally with @panchyo0, but thought I should open up the conversation a bit)_

When version 1.0's tech stack was initially decided on, I proposed we use [Chart.js](https://github.com/chartjs/Chart.js) over [plotly.js](https://github.com/plotly/plotly.js), for performance reasons. At the time, I wrote a couple of codepen examples that refreshed a plot at 33 ms, with four traces with a few thousand points (a somewhat realistic example of what we'd do with CARTA). At the time of creating those pens, plotly.js lagged behind chart.js significantly (taking 17 ms to update the chart, versus 10 ms), so chart.js was chosen as an appropriate charting library for CARTA.

However, since then, both charting libraries have improved performance. chart.js now takes 8 ms to update the chart, and plotly.js now takes 10 ms. However, the big change is that plotly.js now supports WebGL-accelerated charting (using the `scattergl` plot type), and this drops the update time to below 7 ms. 

Plotly.js's WebGL acceleration could enable us to have interactive performance with large scatter plots (catalogue overlays?), and we could also make use of their 3D plots for other uses. We can also fall back on their SVG output when exporting plots.

[Firefly](https://github.com/Caltech-IPAC/firefly) has used plotly.js to great effect.

For now, this issue should just serve as a starting point for the discussion. Do we want to make the migration? How do we handle it? When do we make the migration? (I suggest aiming for 1.4)

**Comment by kswang1029** [comment]
Does chart.js have the potential to perform equal or better than plotly in the future? (this is a hard question I guess)

**Comment by veggiesaurus** [comment]
> Does chart.js have the potential to perform equal or better than plotly in the future? (this is a hard question I guess)

It's hard to say, really. I've clearly been wrong about predicting future performance before (which is why we're using Chart.js now :see_no_evil: )

There are three main graphics rendering approaches that charting libraries can use:
- SVG (vector graphics, very pretty and easily exportable)
- canvas (raster graphics, normally much faster than SVG)
- WebGL (raster graphics, much more complicated than either of the others, but much faster)

We're using WebGL for our contours and tiles, because canvas and SVG would be far too slow (although if we want to _export_ contours nicely, we might need an SVG renderer as well). We use canvas for our AST overlays, because WebGL would be too complicated to implement.

Plotly uses SVG by default, but WebGL if the `scattergl` plot type is used. This is especially important for plotting a large number of points, e.g. catalogue overlays or spectral profiles of 32K cubes . Chart.js uses canvas for all plot types, and as far as I know, has no plans to support WebGL. Both libraries will probably improve in performance over time, but I do think the addition of WebGL to plotly gives it a major advantage. We would still fall back to SVG when exporting images, which would sort of be a best-of-both situation.

**Comment by zarda** [comment]
Is that possible to use both chart.js and plotly.js, which would be on different layers? We could have a WebGL layer of plotly which dedicate to 3D plots. And what if they are not enough in the near future, we can add another layer embedded other library pixi.js or three.js .

**Comment by veggiesaurus** [comment]
It's possible to use both charting libraries, but for consistency, I'd prefer to eventually use one. I think our plan at first is to use plotly for the catalog overlay feature, and then consider migrating the other charts to plotly.

At this point, there's no reason to add pixi or three.js, because our current WebGL layers are at a lower level than those two libraries.

**Comment by veggiesaurus** [comment]
Slight spanner-in-the-works: plotly's latest release has a performance regression in the tests that I was using, and the pre-release of chart.js 3.0 seems to drastically improve performance (3 ms in the test mentioned in the original issue description, down from 8 ms and much lower that plotly in scattergl mode).

not sure if we should consider using plotly only for specific plots in the near future (catalog overlay), or whether it's worth using our own simple WebGL code to plot overlays instead.

**Comment by panchyo0** [comment]
Our code might still having problem with chart.js 3.0. Using our own WebGL code might be better. Then we don't need to rely on other library. if user choose plot catalog in different shape then we need to recalculate the data in backend. Not sure how hard it will be?

**Comment by veggiesaurus** [comment]
> Our code might still having problem with chart.js 3.0. Using our own WebGL code might be better. Then we don't need to rely on other library. if user choose plot catalog in different shape then we need to recalculate the data in backend. Not sure how hard it will be?

no, we'll keep all the data cached in the frontend, and ideally just use the GPU to do the actually translation from data points to rendered shapes. I'm quite happy to handle most of the WebGL stuff so that you can focus on the rest of the catalog features :smile:

**Comment by panchyo0** [comment]
That's smart. Thanks

**Comment by panchyo0** [comment]
Due to the Webgl context limitation (16 from Chrome, 32 configurable with `--max-active-webgl-contexts=32`), we can not update all chart to `scattergl` with Plotlyjs. Translate between `scatter` and `scattergl` will block frontend UI (single thread).
In order to use Webgl for all plots we need to a way to share webgl context between all plots.

https://user-images.githubusercontent.com/22226157/107706245-3ae91e00-6c7d-11eb-9fa0-1ecbcc3c2a5f.mp4

scatter plot 40K Plotlyjs https://codepen.io/panchyo0/pen/YzpwpKQ
scatter plot 40K Chartjs https://codepen.io/panchyo0/pen/wvoMoxb

**Comment by kswang1029** [comment]
> scatter plot 40K Plotlyjs https://codepen.io/panchyo0/pen/YzpwpKQ
> scatter plot 40K Chartjs https://codepen.io/panchyo0/pen/wvoMoxb

in this case, chartjs performs significantly better than plotlyjs. 🤔
